### ISSUE TITLE
Build issue with --minimal (build works, install doesn't)

### ISSUE THREAD

**Issue by gl-yziquel** [Xpra-org/xpra#4182]
**Describe the bug**

Unclear as to how to install from source with --minimal.

**To Reproduce**

I do not expect this issue to be reproducible on just anyone's machine, as this occurs in virtual machine launched and managed by lxd / incus (and the whole point is to get a virtual machine with xpra providing a X layer without having to virtualise more than the bare minimum).

This is a follow up of https://github.com/Xpra-org/xpra/issues/4105 as I come back to it whenever the need is being felt, as it was the case in January and as is again the case in April.

The build instruction `python setup.py build --minimal` works fine. However, this builds and does not install, and I'd wish to have the xpra executable properly installed, and thus usable. The point and problem is that neither `python setup.py install --minimal` nor `python setup.py install` works. Here are some outputs of the build output. (Apologies if they are too verbose or not verbose enough: I don't know how to taylor it to what is necessary w/r to this issue.)

I have made sure that no xpra package is listed in `pip show` before running these commands.

```console
mini-me@virtucon ~/h/c/xpra (master)> python setup.py install --minimal
setup.py install --minimal
Xpra version 6.0
found Cython version 3.1.0a0
found pkg-config version: 1.8.1
ARCH='x86_64'
using INCLUDE_DIRS=['/home/mini-me/.pyenv/versions/3.12.2/include', '/usr/include', '/usr/local/include']
CC='gcc'
CPP='cpp'
enabling 'cython' for 'vsock'
build switches:
* Os                   : No
* PIC                  : Yes
* Xdummy               : Auto
* Xdummy_wrapper       : Auto
* annotate             : No
* argb                 : No
* argb_encoder         : No
* audio                : No
* avif                 : No
* avif_decoder         : No
* avif_encoder         : No
* brotli               : No
* bundle_tests         : No
* client               : No
* clipboard            : No
* codecs               : No
* csc_cython           : No
* csc_libyuv           : No
* cuda_kernels         : No
* cuda_rebuild         : No
* cython               : Yes
* cython_tracing       : No
* cythonize_more       : No
* data                 : No
* dbus                 : No
* debug                : No
* decoders             : No
[...]
* lib_xkbfile          = 1.1.0
* lib_xrandr           = 1.5.2
* lib_xtst             = 1.2.3
Compiling xpra/buffers/membuf.pyx because it changed.
Compiling xpra/buffers/xxh.pyx because it changed.
Compiling xpra/platform/posix/sd_listen.pyx because it changed.
Compiling xpra/net/vsock/vsock.pyx because it changed.
[1/4] Cythonizing xpra/buffers/membuf.pyx
[2/4] Cythonizing xpra/buffers/xxh.pyx
[3/4] Cythonizing xpra/net/vsock/vsock.pyx
[4/4] Cythonizing xpra/platform/posix/sd_listen.pyx
/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
[...]
warning: no previously-included files matching '*.rej' found anywhere in distribution
warning: no previously-included files matching '*.pxi' found anywhere in distribution
no previously-included directories found matching 'build'
no previously-included directories found matching 'dist'
no previously-included directories found matching 'cache'
no previously-included directories found matching 'opt'
install_data_override.run() install_dir=/home/mini-me/.pyenv/versions/3.12.2
install_data_override.run() root_prefix=/home/mini-me/.pyenv/versions/3.12.2
get_conf_dir(/home/mini-me/.pyenv/versions/3.12.2)=/home/mini-me/.pyenv/versions/3.12.2/etc/xpra
convert_templates([]) dirname='fs/etc/xpra', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra'
convert_templates(['conf.d']) dirname='fs/etc/xpra/conf.d', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d'
[...]
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d/42_client_keyboard.conf from 42_client_keyboard.conf.in
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d/50_server_network.conf from 50_server_network.conf.in
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d/55_server_x11.conf from 55_server_x11.conf.in
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d/60_server.conf from 60_server.conf.in
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/conf.d/65_proxy.conf from 65_proxy.conf.in
convert_templates(['content-categories']) dirname='fs/etc/xpra/content-categories', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra/content-categories'
convert_templates(['content-parent']) dirname='fs/etc/xpra/content-parent', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra/content-parent'
convert_templates(['content-type']) dirname='fs/etc/xpra/content-type', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra/content-type'
convert_templates(['http-headers']) dirname='fs/etc/xpra/http-headers', target_dir='/home/mini-me/.pyenv/versions/3.12.2/etc/xpra/http-headers'
generating /home/mini-me/.pyenv/versions/3.12.2/etc/xpra/xpra.conf from xpra.conf.in
  fs/lib/systemd/system/xpra.socket -> /home/mini-me/.pyenv/versions/3.12.2/lib/systemd/system (0o644)
chmod(/home/mini-me/.pyenv/versions/3.12.2/lib/systemd/system/xpra.socket, 0o644)
zip_safe flag not set; analyzing archive contents...
xpra.buffers.__pycache__.membuf.cpython-312: module references __file__
xpra.buffers.__pycache__.xxh.cpython-312: module references __file__
xpra.net.vsock.__pycache__.vsock.cpython-312: module references __file__
xpra.platform.posix.__pycache__.paths.cpython-312: module references __file__
xpra.platform.posix.__pycache__.sd_listen.cpython-312: module references __file__
error: Couldn't find a setup script in /tmp/easy_install-qvk3o6gl/pygobject-3.48.1.tar.gz
```

And now, without `--minimal`:

```console
mini-me@virtucon ~/h/c/xpra (master)> python setup.py install
setup.py install
Xpra version 6.0
found Cython version 3.1.0a0
found pkg-config version: 1.8.1
ARCH='x86_64'
using INCLUDE_DIRS=['/home/mini-me/.pyenv/versions/3.12.2/include', '/usr/include', '/usr/local/include']
CC='gcc'
CPP='cpp'
build switches:
* Os                   : No
* PIC                  : Yes
* Xdummy               : Auto
* Xdummy_wrapper       : Auto
* annotate             : No
* argb                 : Yes
* argb_encoder         : Yes
* audio                : Yes
* avif                 : No
* avif_decoder         : No
* avif_encoder         : No
* brotli               : Yes
* bundle_tests         : No
* client               : Yes
* clipboard            : Yes
* codecs               : Yes
* csc_cython           : Yes
* csc_libyuv           : No
* cuda_kernels         : No
* cuda_rebuild         : Auto
* cython               : Yes
* cython_tracing       : No
* cythonize_more       : No
[...]
Compiling xpra/platform/posix/netdev_query.pyx because it changed.
Compiling xpra/net/vsock/vsock.pyx because it changed.
Compiling xpra/net/lz4/lz4.pyx because it changed.
[ 1/40] Cythonizing xpra/buffers/cyxor.pyx
[ 2/40] Cythonizing xpra/buffers/membuf.pyx
[ 3/40] Cythonizing xpra/buffers/xxh.pyx
[...]
[36/40] Cythonizing xpra/x11/bindings/ximage.pyx
[37/40] Cythonizing xpra/x11/bindings/xwait.pyx
[38/40] Cythonizing xpra/x11/bindings/xwayland.pyx

Error compiling Cython file:
------------------------------------------------------------
...

    cdef Atom xatom(self, str_or_int):
        """Returns the X atom corresponding to the given Python string or Python
        integer (assumed to already be an X atom)."""
        self.context_check("xatom")
        if isinstance(str_or_int, (int, long)):
                                        ^
------------------------------------------------------------

xpra/x11/bindings/core.pyx:92:40: undeclared name not builtin: long
[39/40] Cythonizing xpra/x11/gtk3/bindings.pyx
[40/40] Cythonizing xpra/x11/gtk3/display_source.pyx

Error compiling Cython file:
------------------------------------------------------------
...
        return {"YU12" : "YUV420P", "YV12" : "YVU420P", "GREY" : "YUV420P"}.get(fmt, fmt)

    cdef parse_pixel_format(self, v4l2_format *vid_format):
        if vid_format.fmt.pix.pixelformat==0:
            return ""
        return "".join([chr((vid_format.fmt.pix.pixelformat//(2**(8*x))) % 256) for x in range(4)])
                                                                         ^
------------------------------------------------------------

xpra/codecs/v4l2/virtual.pyx:356:73: Cannot assign type 'double' to 'int'
Traceback (most recent call last):
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1355, in cythonize_one_helper
    return cythonize_one(*m)
           ^^^^^^^^^^^^^^^^^
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1331, in cythonize_one
    raise CompileError(None, pyx_file)
[...]
Cython.Compiler.Errors.CompileError: xpra/x11/bindings/core.pyx
```
I must say I am heavily confused by the build system.

**System Information (please complete the following information):**

This occurs on a 6.5.0-25-generic Linux kernel, Ubuntu Mantic, up to date, installed as an incus (formerly lxd) image as a virtual machine.

The build commit of the xpra repository is 35ada47afd75fff315598ef8a71cd3a10e1d8b7a, 2024-04-02 17:52 +0700. I'll try the same procedure on released commits.

**Comment by totaam** [comment]
> found Cython version 3.1.0a0

That's not a supported version - please try with the stable 3.0.x series.

> `error: Couldn't find a setup script in /tmp/easy_install-qvk3o6gl/pygobject-3.48.1.tar.gz`

Something is wrong with your environment, it looks like something is trying to install `pygobject` - which should already be installed.
The xpra `./setup.py` file does not do that on its own.

> `xpra/x11/bindings/core.pyx:92:40: undeclared name not builtin: long`

Very likely caused by the alpha build of Cython.

> `return "".join([chr((vid_format.fmt.pix.pixelformat//(2**(8*x))) % 256) for x in range(4)])`

Same. Perhaps 0ccc5c8e0afe4aff93b44cf7b396478c3eb42bdb will help. (untested)

**Comment by gl-yziquel** [comment]
Thank you very much. Will try using the released versions (which was what I was going to attempt anyway.)

Hoping to provide some feedback rather soon. Thank you.

**Comment by gl-yziquel** [comment]
The `python setup.py install --minimal` worked on the `v5.0.7` tag of the `xpra` repository. Not on the latest `master` branch. My issue thus seems "temporarily" fixed.

Closing.

Thank you.

**Comment by gl-yziquel** [comment]
OK. So I'll try to follow up. In two parts.

Part 1.

TL;DR: Building xpra 5.0.7 against cython 3.0.10 works, but doesn't work against cython 3.0.x (commit b197288e92f, i.e. the 3.1.0a0 cython version which is on the 3.0.x branch). This seems like a cython regression in gestation.

Instructions used for cython: `python -m build --wheel` followed by `pip install dist/*.whl`.
Instructions used for xpra: `python setup.py build` followed by `python setup.py install`. (I did not use `--minimal` this time.)

Error obtained:

```console
[31/41] Cythonizing xpra/x11/bindings/posix_display_source.pyx                                                         
[32/41] Cythonizing xpra/x11/bindings/randr.pyx                                                                        
                                                                                                                                                                                                                                              
Error compiling Cython file:                                                                                           
------------------------------------------------------------                                                           
...                                                                                                                    
    assert newf>(<int> f), "end of int not found"                                                                      
    cdef unsigned int unewf = newf                                                                                     
    try:                                                   
        n = int(x[f:unewf])                                                                                            
    except (OverflowError, ValueError):                                                                                
        n = long(x[f:unewf])                                                                                                                                                                                                                  
            ^                                              
------------------------------------------------------------                                                           
                                                                                                                                                                                                                                              
xpra/net/bencode/cython_bencode.pyx:48:12: undeclared name not builtin: long                                           
Traceback (most recent call last):                                                                                     
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1247, in cythonize_one_helper                                                                                                   
    return cythonize_one(*m)                                                                                                                                                                                                                  
           ^^^^^^^^^^^^^^^^^                               
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1237, in cythonize_one                                                                                                          
    raise CompileError(None, pyx_file)                                                                                                                                                                                                        
Cython.Compiler.Errors.CompileError: xpra/net/bencode/cython_bencode.pyx                                               
[33/41] Cythonizing xpra/x11/bindings/res.pyx                                                                          
[34/41] Cythonizing xpra/x11/bindings/wait_for_x_server.pyx                                                            
[35/41] Cythonizing xpra/x11/bindings/window.pyx                                                                       
[36/41] Cythonizing xpra/x11/bindings/xi2.pyx                                                                          
[37/41] Cythonizing xpra/x11/bindings/ximage.pyx                                                                                                                                                                                              
[38/41] Cythonizing xpra/x11/bindings/xwait.pyx                                                                        
[39/41] Cythonizing xpra/x11/bindings/xwayland.pyx                                                                     
[40/41] Cythonizing xpra/x11/gtk3/gdk_bindings.pyx                                                                                                                                                                                            
                                                           
Error compiling Cython file:                               
------------------------------------------------------------                                                                                                                                                                                  
...                                                        
        return {"YU12" : "YUV420P", "YV12" : "YVU420P", "GREY" : "YUV420P"}.get(fmt, fmt)                                                                                                                                                     
                                                                                                                                                                                                                                              
    cdef parse_pixel_format(self, v4l2_format *vid_format):                                                            
        if vid_format.fmt.pix.pixelformat==0:                                                                          
            return ""                                      
        return "".join([chr((vid_format.fmt.pix.pixelformat//(2**(8*x))) % 256) for x in range(4)])                                                                                                                                           
                                                                         ^                                             
------------------------------------------------------------                                                           
                                                           
xpra/codecs/v4l2/pusher.pyx:356:73: Cannot assign type 'double' to 'int'                                                                                                                                                                      
[41/41] Cythonizing xpra/x11/gtk3/gdk_display_source.pyx                                                               
Traceback (most recent call last):                                                                                     
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1247, in cythonize_one_helper                                                                                                   
    return cythonize_one(*m)                               
           ^^^^^^^^^^^^^^^^^                               
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1237, in cythonize_one                                                                                                          
    raise CompileError(None, pyx_file)                                                                                 
Cython.Compiler.Errors.CompileError: xpra/codecs/v4l2/pusher.pyx                                                       
                                                                                                                       
Error compiling Cython file:                                                                                           
------------------------------------------------------------                                                           
...                                                        
                else:                                      
                    keysyms = []                           
                    for ks in keysyms_strs:                                                                            
                        if ks in (None, ""):                                                                           
                            keysym = NoSymbol                                                                          
                        elif isinstance(ks, (long, int)):                                                              
                                             ^                                                                         
------------------------------------------------------------                                                           

xpra/x11/bindings/keyboard.pyx:639:45: undeclared name not builtin: long                                               

Error compiling Cython file:                               
------------------------------------------------------------                                                           
...                                                        

    cdef Atom xatom(self, str_or_int):                                                                                 
        """Returns the X atom corresponding to the given Python string or Python                                                                                                                                                              
        integer (assumed to already be an X atom)."""                                                                  
        self.context_check("xatom")                        
        if isinstance(str_or_int, (int, long)):                                                                        
                                        ^                                                                              
------------------------------------------------------------                                                           

xpra/x11/bindings/core.pyx:90:40: undeclared name not builtin: long                                                    
Traceback (most recent call last):                         
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1247, in cythonize_one_helper                                                                                                   
    return cythonize_one(*m)                               
           ^^^^^^^^^^^^^^^^^                               
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1237, in cythonize_one                                                                                                          
    raise CompileError(None, pyx_file)                                                                                 
Cython.Compiler.Errors.CompileError: xpra/x11/bindings/keyboard.pyx                                                    
Traceback (most recent call last):                         
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1247, in cythonize_one_helper                                                                                                   
    return cythonize_one(*m)                               
           ^^^^^^^^^^^^^^^^^                               
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1237, in cythonize_one                                                                                                          
    raise CompileError(None, pyx_file)                                                                                 
Cython.Compiler.Errors.CompileError: xpra/x11/bindings/core.pyx                                                        
multiprocessing.pool.RemoteTraceback:                                                                                  
"""                                                        
Traceback (most recent call last):                         
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/multiprocessing/pool.py", line 125, in worker                                                                                                                                     
    result = (True, func(*args, **kwds))                                                                               
                    ^^^^^^^^^^^^^^^^^^^                                                                                
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/multiprocessing/pool.py", line 48, in mapstar                                                                                                                                     
    return list(map(*args))                                
           ^^^^^^^^^^^^^^^^                                
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1247, in cythonize_one_helper                                                                                                   
    return cythonize_one(*m)                               
           ^^^^^^^^^^^^^^^^^                               
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1237, in cythonize_one                                                                                                          
    raise CompileError(None, pyx_file)                                                                                 
Cython.Compiler.Errors.CompileError: xpra/net/bencode/cython_bencode.pyx                                               
"""                                                        

The above exception was the direct cause of the following exception:                                                   

Traceback (most recent call last):                         
  File "/home/mini-me/home/backyard/hephaestus/xpra-v5.0.7/setup.py", line 2461, in <module>                                                                                                                                                  
    setup_options["ext_modules"] = cythonize(ext_modules,                                                              
                                   ^^^^^^^^^^^^^^^^^^^^^^                                                              
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/site-packages/Cython/Build/Dependencies.py", line 1083, in cythonize                                                                                                              
    result.get(99999)  # seconds                           
    ^^^^^^^^^^^^^^^^^                                      
  File "/home/mini-me/.pyenv/versions/3.12.2/lib/python3.12/multiprocessing/pool.py", line 774, in get                                                                                                                                        
    raise self._value                                      
Cython.Compiler.Errors.CompileError: xpra/net/bencode/cython_bencode.pyx                                               
```

Part 2.a (**without** the --minimal flag):

TL;DR: Building xpra 5.0.7 against cython 3.0.10 works, and building xpra `master` (commit 35ada47afd75fff) against cython 3.0.10 also does work. The build instructions are the same as in part 1.

Part 2.b (**with** the --minimal flag):

TL;DR: Works fine too.

I guess my issue is that I was using a --prefix that wasn't very standard.

So all good. I'll keep this issue closed.

Thank you.

**Comment by totaam** [comment]
> This seems like a cython regression in gestation.

Not so much a regression as a long deprecated feature removal.

> `n = long(x[f:unewf])`
> `xpra/net/bencode/cython_bencode.pyx:48:12: undeclared name not builtin: long`

This code has already been removed from xpra v6.
If you want to run bleeding edge Cython, you should probably also run bleeding edge xpra.

> `return "".join([chr((vid_format.fmt.pix.pixelformat//(2**(8*x))) % 256) for x in range(4)])`
> `xpra/codecs/v4l2/pusher.pyx:356:73: Cannot assign type 'double' to 'int'`

As per above, there is a workaround in git master: 0ccc5c8e0afe4aff93b44cf7b396478c3eb42bdb

> `elif isinstance(ks, (long, int)):`
> `xpra/x11/bindings/keyboard.pyx:639:45: undeclared name not builtin: long`

Also already removed from v6: b9ca743aff5007011f7941c9ea7f2fe54efe26a9

> `if isinstance(str_or_int, (int, long)):`
> `xpra/x11/bindings/core.pyx:90:40: undeclared name not builtin: long`

Also already removed from v6: 52f39eff9b06b1c7ff865874b6aaec1c9448d254

> I guess my issue is that I was using a --prefix that wasn't very standard.

Makes sense.
